### PR TITLE
IGNITE-16414 Fix sorted index spool and sorted aggregate collation planning

### PR DESCRIPTION
**Pull request by alex-plekhanov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/agg/IgniteSortAggregateBase.java [hunk 1]
```diff
@@ -48,7 +47,22 @@
     @Override default Pair<RelTraitSet, List<RelTraitSet>> passThroughCollation(
         RelTraitSet nodeTraits, List<RelTraitSet> inputTraits
     ) {
-        RelCollation collation = RelCollations.of(ImmutableIntList.copyOf(getGroupSet().asList()));
+        RelCollation required = TraitUtils.collation(nodeTraits);
+        ImmutableBitSet requiredKeys = ImmutableBitSet.of(required.getKeys());
+        RelCollation collation;
+
+        if (getGroupSet().contains(requiredKeys)) {
```
Comment (zstan) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/agg/IgniteSortAggregateBase.java:54]:
```suggestion
        if (getGroupSet().contains(requiredKeys)) {
            ImmutableBitSet keysLeft = getGroupSet().except(requiredKeys);
            
            List<RelFieldCollation> newCollationFields = 
                new ArrayList<>(required.getFieldCollations().size() + keysLeft.size());
            
            newCollationFields.addAll(required.getFieldCollations());
```

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/FilterSpoolMergeToSortedIndexSpoolRule.java [hunk 2]
```diff
@@ -75,13 +83,50 @@ private FilterSpoolMergeToSortedIndexSpoolRule(Config cfg) {
...
+        if (inCollation == null || inCollation.isDefault()) {
+            // Create collation by index condition.
+            List<RexNode> lowerBound = idxCond.lowerBound();
+            List<RexNode> upperBound = idxCond.upperBound();
+
+            int cardinality = lowerBound != null ? lowerBound.size() : upperBound.size();
+
+            List<Integer> equalsFields = new ArrayList<>(cardinality);
+            List<Integer> otherFields = new ArrayList<>(cardinality);
+
```
Comment (zstan) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/FilterSpoolMergeToSortedIndexSpoolRule.java:98]: possibly we need to append smt like : 
`            if (lowerBound != null && upperBound != null)
                assert lowerBound.size() == upperBound.size();`

**Review by zstan — APPROVED** [review]
overall looks good

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/agg/IgniteSortAggregateBase.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/agg/IgniteSortAggregateBase.java:54]: What is the difference? `requiredKeys` is a subset of `groupSet`, so size of `required + keysLeft = required + (groupSet - required) = groupSet`

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/FilterSpoolMergeToSortedIndexSpoolRule.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/FilterSpoolMergeToSortedIndexSpoolRule.java:100]: Fixed

**Review by ivandasch — APPROVED** [review]

**Comment by alex-plekhanov** [comment]
Merged to sql-calcite branch.

**Pull request closed without merging**